python/beeai_framework/agents/experimental/agent.py

```diff
@@ -98,42 +102,61 @@ def __init__(
-        run_config = execution or AgentExecutionConfig(
-            max_retries_per_step=3,
-            total_max_retries=3,
-            max_iterations=20,
+    @runnable_entry
+    async def run(self, input: str | list[AnyMessage], /, **kwargs: Unpack[AgentOptions]) -> RequirementAgentOutput:
+        """Execute the agent.
+
+        Args:
+            input: The input to the agent (if list of messages, uses the last message as input)
+            expected_output: Pydantic model or instruction for steering the agent towards an expected output format.
+            total_max_retries: Maximum number of model retries.
+            max_retries_per_step: Maximum number of model retries per step.
+            max_iterations: Maximum number of iterations.
+            backstory: Additional piece of information or background for the agent.
+            signal: The agent abort signal
+            context: A dictionary that can be used to pass additional context to the agent
+
+        Returns:
+            The agent output.
+        """
+        if not input and self._memory.is_empty():
+            raise ValueError(
+                "Invalid input. The input must be a non-empty string or list of messages when memory is empty."
+            )
+        text_input = (
+            input
+            if isinstance(input, str)
+            else (input[-1].text if input and isinstance(input[-1], UserMessage) else "")
+        )
+        run_config = AgentExecutionConfig(
+            max_retries_per_step=kwargs.get("max_retries_per_step", 3),
+            total_max_retries=kwargs.get("total_max_retries", 20),
+            max_iterations=kwargs.get("max_iterations", 10),
         )
+        expected_output = kwargs.get("expected_output")
 
         async def init_state() -> tuple[RequirementAgentRunState, UserMessage | None]:
             state = RequirementAgentRunState(
                 memory=UnconstrainedMemory(), steps=[], iteration=0, answer=None, result=None
             )
             await state.memory.add_many(self.memory.messages)
+            if isinstance(input, list):
+                await state.memory.add_many(input[:-1])
 
             user_message: UserMessage | None = None
-            if prompt:
+            if text_input:
                 task_input = RequirementAgentTaskPromptInput(
-                    prompt=prompt,
-                    context=context,
+                    prompt=text_input,
+                    context=kwargs.get("backstory"),
                     expected_output=expected_output if isinstance(expected_output, str) else None,  # TODO: validate
                 )
                 user_message = UserMessage(self._templates.task.render(task_input))
                 await state.memory.add(user_message)
 
             return state, user_message
 
-        async def handler(run_context: RunContext) -> RequirementAgentRunOutput[TOutput]:
+        async def handler(run_context: RunContext) -> RequirementAgentOutput:
```

Comment:
Now that the run method is async and returns the same type as the handler, is `async def handler` method still needed?

Reply:
No, it's not. We will drop that in a separate PR. Just trying to minimize the amount of potential conflicts with the main branch at this point :)